### ISSUE TITLE
Remote sort, remote paging, column width

### ISSUE THREAD

**Issue by Hacker-CB** [GwtMaterialDesign/gwt-material-table#18]
Hello!

Currently there is not remote sort / remote paging support in the table.

Remote sort: I can take sort context from dataView.getSortContext() and make sorting in the backend side but no sort arrows are displayed in column headers until I will setup sortComparator. I have tried to override isSrotable to return "true", but no affect.

For the paging: DataSource.getDataSize() is called only once and in synced context, so it is impossible to setup total data rows count from the remote request.

Also I found that width of the columns is not static. It can be changed during paging depending on the content width. So it is possible to keep column width persistent during paging?

**Comment by BenDol** [comment]
You can hook into the column sort events for remote sorting functionality, see [here](https://github.com/GwtMaterialDesign/gwt-material-table/blob/master/src/main/java/gwt/material/design/client/ui/table/AbstractDataTable.java#L807).

Static columns widths is something I will enable tonight. Probably the best way to achieve this will be by adding Column.setStyleProperty("width", "100px") for example.

@kevzlou7979 Can you look into the pager issue?

**Comment by BenDol** [comment]
You can also see the InfiniteDataView how it handles the sorting is like what you are expecting. https://github.com/GwtMaterialDesign/gwt-material-table/blob/master/src/main/java/gwt/material/design/client/data/infinite/InfiniteDataView.java#L291

**Comment by BenDol** [comment]
Can make your header width static now: https://github.com/GwtMaterialDesign/gwt-material-table/commit/37a6dc908608172c7d0d28d0d76ae4a868f55ee8

**Comment by kevzlou7979** [comment]
Cool

**Comment by Hacker-CB** [comment]
> You can also see the InfiniteDataView how it handles the sorting is like what you are expecting. https://github.com/GwtMaterialDesign/gwt-material-table/blob/master/src/main/java/gwt/material/design/client/data/infinite/InfiniteDataView.java#L291

Seems that I need to implement doSort() method to make remote sorting.
I have next ideas to do that:
1. Make something like MaterialRemoteDataTable + RemoteDataView
2. Make just RemoteDataView and make avaibility to use it in MaterialDataTable?
3. Add "bool remoteSort" property to AbstractDataView and just do nothing inside doSort() method? And call pager.refresh() inside table.addSortColumnHandler()?

What about ideas for multiply columns sorting (List<SortContext> instead of just one SortContext)?

**Comment by Hacker-CB** [comment]
I have added pull request to allow remote sorting in any DataView: https://github.com/GwtMaterialDesign/gwt-material-table/pull/20

**Comment by Hacker-CB** [comment]
> For the paging: DataSource.getDataSize() is called only once and in synced context, so it is impossible to setup total data rows count from the remote request.

I see only one solution to fix this, we need to call MaterialDataPager::setTotalRows() and refresh all paging widgets after each load request. So, actually we need to add async callback to DataSource::load() method. In this case totalRows will be updated after each load() request.

For example, Sencha GXT has similar loader:
```
/**
 * Defines the interface for objects that can retrieve data.
 *  
 * @param <C> the type of data used to configure the load from the proxy
 * @param <D> the type of data being returned by the data proxy
 */
public interface DataProxy<C, D> {

  /**
   * Data should be retrieved using the specified load config. When specified,
   * the <code>DataReader</code> can be used to "process" the raw data.
   * 
   * @param loadConfig the load config object to be passed to server
   * @param callback the data callback
   */
  public void load(C loadConfig, Callback<D, Throwable> callback);

}
```
I think it is better to change DataSource.load() signature to callback now. It will be much simple to use any async data loaders in future.

**Comment by remarkNS** [comment]
+1 for callback

**Comment by Hacker-CB** [comment]
I have fully implemented async loading. Pull request: https://github.com/GwtMaterialDesign/gwt-material-table/pull/21
I have tested it in table with pagination, but not tested with infinite table.

Here is list of mayor modifications:
- InfiniteDataSource no longer needed. DataSource now can be used for all common cases.
- MaterialPager can use remote sort out-of-the-box by subscribing to table.addSortColumnHandler() from MaterialPager.onLoad() method
- MaterialPager use provided total records count for pagination

**Comment by BenDol** [comment]
So far I have tested the changes and make some modifications but it seems the pager loading doesn't work in the demo anymore.

**Comment by Hacker-CB** [comment]
Ok, I will try to make it work today.

**Comment by kevzlou7979** [comment]
@Hacker-CB Also if you have a chance, please update also the demo (Addins - DataTable) with the recent changes. Thanks in advance

**Comment by Hacker-CB** [comment]
I found that raw data loaders for dataview and table expect startIndex starting from "1", not from "0". Why it is so strange design?
```
DataView.loaded(int startIndex, List<T> data)
AbstractDataTable.loaded(int startIndex, List<T> data)
```
I have implemented DataSource that first record is started from "0".

I will make some changes and will push all of them later today.

**Comment by Hacker-CB** [comment]
I think that better way is to change all data design to index starting from "0". It is more standard.
@BenDol , @kevzlou7979  - how do you think?

**Comment by Hacker-CB** [comment]
I made two pool requests. Seems that now all works fine.

- https://github.com/GwtMaterialDesign/gwt-material-table/pull/23
- https://github.com/GwtMaterialDesign/gwt-material-table-sample/pull/1

**Comment by kevzlou7979** [comment]
Great Job 👍

**Comment by BenDol** [comment]
@Hacker-CB not actually the intended functionality, just tested that with the InfiniteDataView and it calls `loaded(...)` with startIndex (offset) as 0, infact it requires it to be a 0 based value. Is it an issue with the `ListDataSource` or perhaps the `MaterialDataPager`?

**Comment by BenDol** [comment]
Seems to be an issue with the `MaterialDataPager` returning the startIndex as '1' rather than '0'. The whole MaterialDataPager is really messy, needs to be rewritten to be completely honest.

**Comment by Hacker-CB** [comment]
@BenDol , you are right, MaterialDataPager was not written good. So, what about https://github.com/GwtMaterialDesign/gwt-material-table/pull/23 as a temporary fix to move forward?
